File: frontend/apps/ios/Pantopus/Features/CeremonialMailOpen/CeremonialMailOpenViewModel.swift
//
//  CeremonialMailOpenViewModel.swift
//  Pantopus
//
//  Fetches the mail item via `GET /api/mailbox/v2/item/:id`, projects
//  it into a `CeremonialMailLetter`, owns the four-phase progression
//  (sealed → breaking → open → replying), and exposes a play/pause
//  hook for the optional voice postscript.
//

import Foundation
import Observation

@Observable
@MainActor
public final class CeremonialMailOpenViewModel {
    public private(set) var state: CeremonialMailOpenState = .loading
    public private(set) var isVoicePlaying: Bool = false

    private let mailId: String
    private let api: APIClient

    public init(mailId: String, api: APIClient = .shared) {
        self.mailId = mailId
        self.api = api
    }

    public func load() async {
        state = .loading
        do {
            let response: MailboxV2ItemResponse =
                try await api.request(MailboxV2Endpoints.item(mailId: mailId))
            let letter = Self.project(response: response, mailId: mailId)
            state = .loaded(letter, phase: .sealed)
        } catch {
            let message = (error as? APIError)?.errorDescription ?? "Couldn't load this letter."
            state = .error(message: message)
        }
    }

    public var phase: CeremonialMailPhase {
        if case let .loaded(_, phase) = state { return phase }
        return .sealed
    }

    public var letter: CeremonialMailLetter? {
        if case let .loaded(letter, _) = state { return letter }
        return nil
    }

    /// Step the seal-break ceremony forward. View calls this once
    /// the user taps the envelope; the breaking phase animates and
    /// then transitions to `.open` after `~600ms`.
    public func startBreakingSeal() async {
        guard case let .loaded(letter, phase) = state, phase == .sealed else { return }
        state = .loaded(letter, phase: .breaking)
        try? await Task.sleep(nanoseconds: 600_000_000)
        if case .loaded = state {
            state = .loaded(letter, phase: .open)
        }
    }

    public func openImmediately() {
        guard case let .loaded(letter, _) = state else { return }
        state = .loaded(letter, phase: .open)
    }

    public func enterReplying() {
        guard case let .loaded(letter, _) = state else { return }
        state = .loaded(letter, phase: .replying)
    }

    public func resetToOpen() {
        guard case let .loaded(letter, _) = state else { return }
        state = .loaded(letter, phase: .open)
    }

    public func toggleVoicePlayback() {
        isVoicePlaying.toggle()
    }

    public func stopVoicePlayback() {
        isVoicePlaying = false
    }

    // MARK: - Projection

    static func project(
        response: MailboxV2ItemResponse,
        mailId: String
    ) -> CeremonialMailLetter {
        let item = response.mail
        let payload = item.objectPayload?.objectValue ?? [:]
        let stationery = CeremonialMailStationeryTone(
            wire: payload["stationeryTheme"]?.stringValue
        )
        let ink = CeremonialMailInkTone(wire: payload["inkSelection"]?.stringValue)
        let seal = CeremonialMailSealTone(wire: payload["sealChoice"]?.stringValue)
        let voiceUri = payload["voicePostscriptUri"]?.stringValue
        let body = item.base.content ?? ""
        let paragraphs = body
            .components(separatedBy: "\n\n")
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
        let sender = CeremonialSenderCard(
            displayName: item.senderDisplay,
            handle: item.sender?.username,
            trustLabel: trustLabel(for: item.senderTrust),
            avatarUrl: nil
        )
        let category = "letter"
        return CeremonialMailLetter(
            mailId: mailId,
            sender: sender,
            category: category,
            subject: item.base.subject ?? "A letter",
            bodyParagraphs: paragraphs.isEmpty ? [body] : paragraphs,
            stationery: stationery,
            ink: ink,
            seal: seal,
            voicePostscriptUri: voiceUri,
            receivedAt: item.base.createdAt,
            outcomeCtas: CeremonialMailLetter.defaultOutcomeCtas()
        )
    }

    static func trustLabel(for raw: String?) -> String? {
        switch raw {
        case "verified_gov", "verified_utility", "verified_business": "Verified"
        case "pantopus_user": "Pantopus friend"
        case "partial": "Partial trust"
        case "none", nil: nil
        default: nil
        }
    }
}

// MARK: - JSONValue accessors

private extension JSONValue {
    var objectValue: [String: JSONValue]? {
        if case let .object(value) = self { return value }
        return nil
    }

    var stringValue: String? {
        if case let .string(value) = self { return value }
        return nil
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/ceremonial_mail_open/CeremonialMailOpenViewModel.kt
@file:Suppress("PackageNaming", "LongMethod", "MagicNumber")

package app.pantopus.android.ui.screens.ceremonial_mail_open

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.mailbox.MailboxRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import org.json.JSONObject
import javax.inject.Inject

@HiltViewModel
class CeremonialMailOpenViewModel
    @Inject
    constructor(
        private val repository: MailboxRepository,
        savedStateHandle: SavedStateHandle,
    ) : ViewModel() {
        private val mailId: String = savedStateHandle.get<String>(MAIL_ID_KEY) ?: ""

        private val _state = MutableStateFlow<CeremonialMailOpenUiState>(CeremonialMailOpenUiState.Loading)
        val state: StateFlow<CeremonialMailOpenUiState> = _state.asStateFlow()

        private val _isVoicePlaying = MutableStateFlow(false)
        val isVoicePlaying: StateFlow<Boolean> = _isVoicePlaying.asStateFlow()

        fun load() {
            if (mailId.isBlank()) {
                _state.value = CeremonialMailOpenUiState.Error("Missing mail id.")
                return
            }
            _state.value = CeremonialMailOpenUiState.Loading
            viewModelScope.launch {
                when (val result = repository.item(mailId)) {
                    is NetworkResult.Success -> {
                        val letter = project(result.data, mailId)
                        _state.value = CeremonialMailOpenUiState.Loaded(letter, CeremonialMailPhase.Sealed)
                    }
                    is NetworkResult.Failure ->
                        _state.value = CeremonialMailOpenUiState.Error("Couldn't load this letter.")
                }
            }
        }

        /** Step the seal-break ceremony forward. The view triggers
         *  this when the user taps the envelope. */
        fun startBreakingSeal() {
            val current = _state.value as? CeremonialMailOpenUiState.Loaded ?: return
            if (current.phase != CeremonialMailPhase.Sealed) return
            _state.value = current.copy(phase = CeremonialMailPhase.Breaking)
            viewModelScope.launch {
                delay(600)
                val now = _state.value as? CeremonialMailOpenUiState.Loaded ?: return@launch
                if (now.phase == CeremonialMailPhase.Breaking) {
                    _state.value = now.copy(phase = CeremonialMailPhase.Open)
                }
            }
        }

        fun openImmediately() {
            val current = _state.value as? CeremonialMailOpenUiState.Loaded ?: return
            _state.value = current.copy(phase = CeremonialMailPhase.Open)
        }

        fun enterReplying() {
            val current = _state.value as? CeremonialMailOpenUiState.Loaded ?: return
            _state.value = current.copy(phase = CeremonialMailPhase.Replying)
        }

        fun resetToOpen() {
            val current = _state.value as? CeremonialMailOpenUiState.Loaded ?: return
            _state.value = current.copy(phase = CeremonialMailPhase.Open)
        }

        fun toggleVoicePlayback() {
            _isVoicePlaying.value = !_isVoicePlaying.value
        }

        fun stopVoicePlayback() {
            _isVoicePlaying.value = false
        }

        companion object {
            const val MAIL_ID_KEY = "mailId"

            internal fun project(
                response: app.pantopus.android.data.api.models.mailbox.v2.MailboxV2ItemResponse,
                mailId: String,
            ): CeremonialMailLetter {
                val item = response.mail
                // The backend's `object_payload` arrives as an opaque
                // map — pull the ceremonial slots by key. The wider
                // payload is intentionally untyped at the repo layer
                // because each mail_type has a different shape.
                val payload =
                    runCatching {
                        item.objectPayload?.let { JSONObject(it.toString()) }
                    }.getOrNull()
                val stationery =
                    CeremonialMailStationeryTone.fromWire(payload?.optString("stationeryTheme")?.takeIf { it.isNotEmpty() })
                val ink = CeremonialMailInkTone.fromWire(payload?.optString("inkSelection")?.takeIf { it.isNotEmpty() })
                val seal = CeremonialMailSealTone.fromWire(payload?.optString("sealChoice")?.takeIf { it.isNotEmpty() })
                val voiceUri = payload?.optString("voicePostscriptUri")?.takeIf { it.isNotEmpty() }
                val body = item.content.orEmpty()
                val paragraphs =
                    body.split("\n\n")
                        .map { it.trim() }
                        .filter { it.isNotEmpty() }
                val sender =
                    CeremonialSenderCard(
                        displayName = item.senderDisplay ?: item.sender?.name ?: "Someone",
                        handle = item.sender?.username,
                        trustLabel = trustLabel(item.senderTrust),
                        avatarUrl = null,
                    )
                return CeremonialMailLetter(
                    mailId = mailId,
                    sender = sender,
                    category = "letter",
                    subject = item.subject ?: "A letter",
                    bodyParagraphs = if (paragraphs.isEmpty()) listOf(body) else paragraphs,
                    stationery = stationery,
                    ink = ink,
                    seal = seal,
                    voicePostscriptUri = voiceUri,
                    receivedAt = item.createdAt,
                    outcomeCtas = CeremonialMailLetter.defaultOutcomeCtas(),
                )
            }

            internal fun trustLabel(raw: String?): String? =
                when (raw) {
                    "verified_gov", "verified_utility", "verified_business" -> "Verified"
                    "pantopus_user" -> "Pantopus friend"
                    "partial" -> "Partial trust"
                    else -> null
                }
        }
    }
